[quant][graphmode] quant_fusion for all methods that is called

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#26078 [quant][graphmode] quant_fusion for all methods that is called**
* #25974 [quant][graphmode] use whitelist for selecting observed values
* #25625 [quant][graphmode] Fold quantize op into module
* #25624 [quant][graphmode] Add fusion for quantized linear
* #25623 Port fuse_linear from pytorch/tvm
* #25505 [quant][graphmode] Support quantizing any methods called
* #25504 [quant][graphmode] Skip inserting duplicate observers

Summary:
Add a QuantFusion to recursively fuse all the graphs that is invoked by
the method

Test Plan:
python test/test_jit.py 'TestJit.test_quant_fusion_module'

Reviewers:
pt1quant
Subscribers:

Tasks:

Tags:

Differential Revision: [D17348985](https://our.internmc.facebook.com/intern/diff/D17348985)

jerryzh168 committed:
[quant][graphmode] quant_fusion for all methods that is called

Summary:
Add a QuantFusion to recursively fuse all the graphs that is invoked by
the method

Test Plan:
python test/test_jit.py 'TestJit.test_quant_fusion_module'

Reviewers:
pt1quant
Subscribers:

Tasks:

Tags:

ZolotukhinM commented:
In the end the fusion will happen in one of two places:
- in graph executor along with other optimizations: at that point we won't have modules (and in fact we won't need them) - operating on a graph scope would work just fine as it is something we would execute.
- in mobile workflow before serialization - in that case we want to run fusion in advance, but I think we would like to run them everywhere, not just in the method called by other methods - that would help us to use less operators in the end and potentially we could reduce the set of operators we need to deploy to mobile.

In neither of these scenarios we would be walking call-methods to find graphs to optimize, so I don't think this PR is the right approach.

jerryzh168 committed:
Update on "[quant][graphmode] quant_fusion for all methods that is called"


Summary:
Add a QuantFusion to recursively fuse all the graphs that is invoked by
the method

Test Plan:
python test/test_jit.py 'TestJit.test_quant_fusion_module'

Reviewers:
pt1quant
Subscribers:

Tasks:

Tags:
